### PR TITLE
nixos/acme: Backport account rate limit fixes and tmpfile removal

### PR DESCRIPTION
**Pull request by m1cr0man**
###### Motivation for this change

Backport of #106857 (excluding the merge commit, of course)

It [has been shown](https://github.com/NixOS/nixpkgs/issues/101445#issuecomment-771243252) to fix/prevent JWS issues pretty reliably at this point, and there has been demand to get these features backported.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pjones** [comment]
@m1cr0man What impact will this have on existing NixOS 20.09 users with ACME certs?

File: nixos/modules/security/acme.nix [hunk 1]
```diff
@@ -7,6 +7,11 @@ let
   numCerts = length (builtins.attrNames cfg.certs);
   _24hSecs = 60 * 60 * 24;
 
+  # Used to make unique paths for each cert/account config set
+  mkHash = with builtins; val: substring 0 20 (hashString "sha256" val);
```
Comment (pjones) [inline nixos/modules/security/acme.nix:11]: @m1cr0man What happens if the first 20 characters of the hash are not unique?

**Comment by matthiasbeyer** [comment]
I'm hitting this error, what keeps that PR from being merged?

File: nixos/modules/security/acme.nix (hunk 1 above)
Comment (m1cr0man) [reply nixos/modules/security/acme.nix:11]: Cert/account folders have the same name for different certs. Lego creates subdirectories that are unique per domain/account. This is a non issue.

**Comment by m1cr0man** [comment]
> @m1cr0man What impact will this have on existing NixOS 20.09 users with ACME certs?

Hash functions haven't actually changed since [this PR](https://github.com/NixOS/nixpkgs/pull/107769/files#diff-0e236440724996cc4ac28adc872acb41d4c0a235e9c1eff679618ba9f45918feR107), so nothing.


> I'm hitting this error, what keeps that PR from being merged?

Just needs to be noticed by maintainers that can merge it I suppose. My PRs always take a while to get in, I'm not sure what the protocol is around bumping them.

**Comment by matthiasbeyer** [comment]
> I'm not sure what the protocol is around bumping them.

There is none. Maybe @primeos can have a look here? :small_airplane:

**Comment by primeos** [comment]
> Maybe @primeos can have a look here?

I'm not using this module so unfortunately I cannot properly review this PR.

> Just needs to be noticed by maintainers that can merge it I suppose.

IMO pinging @NixOS/acme seems like a good idea (3 team members are Nixpkgs committers) and I'd also drop a comment on the PR for master (i.e. #106857) so that the original reviewers get a notification about the backport (a comment because only mentioning a PR creates a link in the web UI but doesn't generate a notification). Hope that helps.

**Comment by matthiasbeyer** [comment]
On 10-02-2021 05:03:45, Michael Weiss wrote:
> Hope that helps.

Thanks!

**Comment by symphorien** [comment]
@GrahamcOfBorg test acme

**Comment by yoctocell** [comment]
I just bumped into this issue today.  Tried this PR and everything is working again.  Thanks!

**Comment by aanderse** [comment]
Anyone wanna merge this? I vote yes.

**Comment by symphorien** [comment]
This was somehow a breaking change:
```
Feb 14 04:44:24 blitiri systemd[1]: Starting Renew ACME certificate for bitwarden.xlumurb.eu...
Feb 14 04:44:24 blitiri acme-bitwarden.xlumurb.eu-start[16084]: + mkdir -p /var/lib/acme/acme-challenge/.well-known/acme-challenge
Feb 14 04:44:24 blitiri acme-bitwarden.xlumurb.eu-start[16084]: + chown acme:nginx /var/lib/acme/acme-challenge/.well-known /var/lib/acme/acme-challenge/.well-known/acme-challenge
Feb 14 04:44:24 blitiri systemd[1]: acme-bitwarden.xlumurb.eu.service: Main process exited, code=exited, status=1/FAILURE
Feb 14 04:44:25 blitiri acme-bitwarden.xlumurb.eu-start[16086]: chown: changing ownership of '/var/lib/acme/acme-challenge/.well-known': Operation not permitted
Feb 14 04:44:24 blitiri systemd[1]: acme-bitwarden.xlumurb.eu.service: Failed with result 'exit-code'.
Feb 14 04:44:24 blitiri systemd[1]: Failed to start Renew ACME certificate for bitwarden.xlumurb.eu.
```


```
#  tree -augp /var/lib/acme/acme-challenge/
/var/lib/acme/acme-challenge/
└── [drwxr-xr-x root     root    ]  .well-known
    └── [drwxr-xr-x acme     nginx   ]  acme-challenge
```
Removing /var/lib/acme/acme-challenge/ does not help.

**Comment by symphorien** [comment]
Also tried `systemd-tmpfiles --create` and wiping `/var/lib/acme` without success.

**Comment by m1cr0man** [comment]
I can't see from the Acme module how you would ever end up with a .well-known folder owned by root. All the services run as acme user and whatever configured group (nginx in your case). Even without the chown, this would cause lego to fail down the line. Can you check for systemd tmpfiles rules associated with that directory? As of this PR, there are none from the Acme module. You can use this command: `systemd-tmpfiles --cat-config | grep .well-known`

**Comment by aanderse** [comment]
If there is a `tmpfile` rule for something inside `.well-known` but `.well-known` doesn't exist then `.well-known` will automatically be created and owned by `root`. Not sure if relevant, just thought I would mention because many people don't realize this. 🤷‍♂️

**Comment by symphorien** [comment]
Wiping `/var/lib/acme` and then rebooting fixes the problem. This suggests that the new module is correct when starting from empty state, but that the previous module could end up in situations which are incompatible with the new module.

`systemd-tmpfiles --cat-config | grep .well-known` has no output on my system.

**Comment by m1cr0man** [comment]
> Wiping `/var/lib/acme` and then rebooting fixes the problem. This suggests that the new module is correct when starting from empty state, but that the previous module could end up in situations which are incompatible with the new module.
> 
This too is odd because you deleted your `/var/lib/acme` directory. I can't think of anything that would persist like that before and after the patch...

**Comment by symphorien** [comment]
There is the systemd state: maybe some already started .target or .service with RemainAfterExit was not restarted when I ran `nixos-rebuild switch` to retry the failed upgrade.

**Comment by m1cr0man** [comment]
The only service using RemainAfterExit is acme-fixperms.service, which does not interact with webroot at all. There hasn't been any other services using that flag for a few months now, possibly never since the 20.09 release. There aren't (even before this PR, too) any references to paths under .well-known/ in any StateDirectory options either. My best guess right now is that systemd-tmpfiles did not correctly reload after you did nixos-rebuild and thus was trying to create the path. As per what @aanderse explained, and looking at the tmpfiles config before this PR, it would make sense (taken from my own system):
```
systemd-tmpfiles --cat-config | grep .well-known
d /var/lib/acme/acme-challenges/.well-known/acme-challenge - acme acme
```

**Comment by jbgi** [comment]
We hit this [issue](https://github.com/NixOS/nixpkgs/pull/112145#issuecomment-778777656) too while upgrading to 20.09 (through nixops). Moving `/var/lib/acme` and reboot did work, but still inconvenient.

Also I tried wiping `/var/lib/acme` again after the reboot, and the acme service failed with same error.

**Pull request merged**